test/ibmq/test_ibmq_job_attributes.py

```diff
@@ -106,6 +106,50 @@ def test_job_name(self, provider):
         self.assertEqual(len(retrieved_jobs), 1)
         self.assertEqual(job_id, retrieved_jobs[0].job_id())
 
+    @requires_provider
+    def test_job_name_change(self, provider):
+        """Test changing the name associated with a job."""
+        backend = provider.get_backend('ibmq_qasm_simulator')
+        qobj = assemble(transpile(self._qc, backend=backend), backend=backend)
+
+        # Use a unique job name
+        initial_job_name = str(time.time()).replace('.', '')
+        job = backend.run(qobj, job_name=initial_job_name, validate_qobj=True)
+        job_id = job.job_id()
+        # TODO No need to wait for job to run once api is fixed
+        while job.status() not in JOB_FINAL_STATES + (JobStatus.RUNNING,):
+            time.sleep(0.5)
+        rjob = provider.backends.retrieve_job(job_id)
+        self.assertEqual(rjob.name(), initial_job_name)
+
+        new_names_to_test = [
+            '',  # empty string as name.
+            '{}_new'.format(str(time.time()).replace('.', ''))  # non-empty name.
+        ]
+        for new_name in new_names_to_test:
+            with self.subTest(new_name=new_name):
+                # Update the job name.
+                update_successful = job.change_name(new_name)
+                self.assertTrue(update_successful,
+                                'Updating the name for job {} from "{}" to "{}" '
+                                'was unsuccessful.'.format(job_id, job.name(), new_name))
+
+                # Check retrieving the job by its new name using partial matching.
+                job_name_partial = new_name[8:]
+                retrieved_jobs = provider.backends.jobs(backend_name=backend.name(),
+                                                        job_name=job_name_partial)
+                self.assertGreaterEqual(len(retrieved_jobs), 1)
+                retrieved_job_ids = {job.job_id() for job in retrieved_jobs}
+                self.assertIn(job_id, retrieved_job_ids)
+
+                if new_name:
+                    # Check retrieving the job by its new name using regular expressions.
+                    job_name_regex = '^{}$'.format(new_name)
+                    retrieved_jobs = provider.backends.jobs(backend_name=backend.name(),
+                                                            job_name=job_name_regex)
+                    self.assertEqual(len(retrieved_jobs), 1)
+                    self.assertEqual(job_id, retrieved_jobs[0].job_id())
```

Comment:
I don't think testing retrieving the new name using partial/regex is necessary as they're covered by other tests. You should, however, do a `job.refresh()` (which will re-fetch job attributes) and verify the job name is the new one.

Reply:
I could see how they could be redundant. The idea was more so to test that it was possible to retrieve the job after updating the name (ensuring the new name was changed successfully in the server). I'll go ahead and replace that with your suggested test, however, since its essentially the same idea.